### PR TITLE
Add the `ValueUpdater` widget

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This PR adds a new widget, `ValueUpdater`, which allows for toggling between values and rebuilding a widget with the new value.

The `ValueUpdater` widget is useful for triggering changes in other widgets without having to create a stateful widget. For example, it can be used to trigger an animation when a widget is shown for the first time, or to change the value of a widget when a button is pressed.

The `ValueUpdater` constructor takes a builder function, which takes the current value and returns a widget. It also takes an `initialValue`, which is the value to use when the widget is first built. It also takes a `valueNotifier` which will switch from the `initialValue` and the value of the `valueNotifier.value` on the first frame and then to the new value whenever the `valueNotifier` notifies of a change.

Optionally, There is a `ValueNotifier.once` constructor that takes a `value`. The widget will switch from the `initialValue` to the `value` on the frame after the widget is first built, but remain at `value` after that. This allows easy implicit animation without having to trigger the animation from a `StatefulWidget`.

Adds the widget and two examples.

## Tests
 - Added tests for the widget, as well as for the examples.

### CODE REVIEW

File: examples/api/lib/widgets/value_updater/value_updater.1.dart [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            child: ValueUpdater<AlignmentGeometry>(
+              initialValue: Alignment.center,
+              valueNotifier: _notifier,
+              builder: (BuildContext context, AlignmentGeometry value) {
+                return AnimatedAlign(
+                  alignment: value,
+                  duration: const Duration(seconds: 1),
+                  child: const Padding(
+                    padding: EdgeInsets.all(8),
+                    child: FlutterLogo(size: 100.0),
```
Comment (goderbauer) [inline examples/api/lib/widgets/value_updater/value_updater.1.dart:43]: Should the ValueUpdater take an optional child parameter that is passed through to the builder? Then, if you have a complex child it doesn't need to rebuild when the builder needs to update (we do this for some other widgets as well, e.g. https://main-api.flutter.dev/flutter/widgets/AnimatedBuilder-class.html (see "performance considerations").

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'framework.dart';
+
+/// A builder function used by the [ValueUpdater] widget that takes the current
+/// value and returns a widget.
+typedef ValueUpdaterBuilder<T> = Widget Function(BuildContext context, T value);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:8]: Could reuse `ValueWidgetBuilder<T>` if we add the optional child argument.

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+/// A widget that updates a value for the widgets built using a builder.
+///
+/// The [ValueUpdater] widget can be used to avoid the need for creating a
+/// stateful widget just to trigger a change in another widget on creation.
+///
+/// If not given a [valueNotifier], will start at the [initialValue] (or the
+/// [initialValue], if given) and toggle to the [value] on the frame after the
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:16]: > will start at the [initialValue] (or the [initialValue], if given)

It mentions `initialValue` twice?

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+/// done without needing to create a stateful widget.
+///
+/// ** See code in examples/api/lib/widgets/value_updater/value_updater.0.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to use the [ValueUpdater] widget to trigger an
+/// [AnimatedAlign] animation when a widget is shown for the first time, and
+/// then change the value of the alignment and animate each time a button is
+/// pressed. This is done without needing to create a stateful widget.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:39]: I think this example actually needs a stateful widget (or some other stateful thing to hold the Notifier), see comment in example.

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+///
+/// ** See code in examples/api/lib/widgets/value_updater/value_updater.1.dart **
+/// {@end-tool}
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:43]: nit: remove blank line

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+class ValueUpdater<T> extends StatefulWidget {
+  /// Creates a [ValueUpdater] widget that updates the value based on
+  /// [valueNotifier] changes.
+  ///
+  /// The [builder], [initialValue], and [valueNotifier] are
+  /// required.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:49]: nit: this doesn't match the signature below where valueNotifier seems to be optional?

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    super.key,
+    required this.builder,
+    required this.initialValue,
+    this.valueNotifier,
+  })  : value = null;
+
+  /// Creates a [ValueUpdater] widget that updates the value only once (the
+  /// first time it is built) based on [value].
+  ///
+  /// The [builder], [initialValue], and [value] are required.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:60]: Same here, but with `value`.

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  })  : valueNotifier = null;
+
+  /// A builder function which takes a [BuildContext], and the current [value]
+  /// (or the value of the [valueNotifier]) and returns a widget.
+  ///
+  /// Required.
+  final ValueUpdaterBuilder<T> builder;
+
+  /// The value to pass to the [builder] when the widget is first built.
+  ///
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:75]: Maybe explicitly document: what happens if you rebuild the widget with a different initialValue after the widget has been build? I assume this is just ignored?

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+  /// The optional value to pass to the [builder] when the widget is built the
+  /// second and subsequent times.
+  ///
+  /// Ignored if [valueNotifier] is set.
+  ///
+  /// If not specified, then the current value of [valueNotifier] is used. One
+  /// of [value] or [valueNotifier] must be set.
+  ///
+  /// Only one of [valueNotifier] or [value] may be set at a time, but one of
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:87]: nit: this paragraph seems to just repeat what the previous one was saying?

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+  /// The value to pass to the [builder] when the widget is first built.
+  ///
+  /// Required.
+  final T initialValue;
+
+  /// The optional value to pass to the [builder] when the widget is built the
+  /// second and subsequent times.
+  ///
+  /// Ignored if [valueNotifier] is set.
+  ///
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:83]: Maybe document what happens if the widget rebuilds with a different `value`? I'd assume it just passes that new value to the builder?

File: examples/api/lib/widgets/value_updater/value_updater.1.dart [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    return MaterialApp(
+      home: ToggleValueWithNotifierExample(),
+    );
+  }
+}
+
+class ToggleValueWithNotifierExample extends StatelessWidget {
+  ToggleValueWithNotifierExample({super.key});
+
+  final ValueNotifier<AlignmentGeometry> _notifier = ValueNotifier<AlignmentGeometry>(Alignment.bottomLeft);
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply examples/api/lib/widgets/value_updater/value_updater.1.dart:25]: Also, _notifier should get disposed when its life is done.

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    super.initState();
+    _value = widget.value ?? widget.initialValue;
+    SchedulerBinding.instance.addPostFrameCallback((_) {
+      setState(() {
+        _value = widget.valueNotifier?.value ?? widget.value ?? _value;
+        _triggered = switch (_triggered) {
+          _ValueUpdaterStatus.off => _ValueUpdaterStatus.on,
+          _ValueUpdaterStatus.on => _ValueUpdaterStatus.off,
+          _ValueUpdaterStatus.initial => _ValueUpdaterStatus.on,
+        };
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:126]: Can _triggered ever reach the `off` state? Seems like it starts out as `initial` and then this post frame callback triggers once to move it to `on`, but it appears there is no logic to move it to `off`?

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return widget.builder(
+      context,
+      switch (_triggered) {
+        _ValueUpdaterStatus.off => widget.value != null ? widget.initialValue : _value,
+        _ValueUpdaterStatus.on => widget.value ?? _value,
+        _ValueUpdaterStatus.initial => widget.initialValue,
+      },
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:168]: Wondering: Do we even need to keep track of _triggered? If `_value` were to be assigned `widget.initialValue` in `initState` could we just always pass `_value` here and remove _ValueUpdaterStatus altogether?

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+  void didUpdateWidget(covariant ValueUpdater<T> oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.valueNotifier != widget.valueNotifier) {
+      oldWidget.valueNotifier?.removeListener(_trigger);
+      widget.valueNotifier?.addListener(_trigger);
+      _value = widget.valueNotifier?.value ?? widget.value ?? _value;
+    }
+    if (oldWidget.value != widget.value) {
+      final T? value = widget.value;
+      if (value != null) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:148]: Here and in the postframecallback is a hidden limitation that you can never use `null` as a target value. Is that intentional? On the surface, I would expect you should be able to do something like `ValueUpdater<int?>(initialValue: 10, value: null)`? I think in the current setup, the null would just get ignored? If this is an intentional limitation, let's document it. And either way, maybe we should add some tests for the null case?

File: packages/flutter/test/widgets/value_updater_test.dart [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+      ),
+    );
+
+    expect(find.text('0'), findsOneWidget);
+    await tester.pump();
+    expect(find.text('1'), findsOneWidget);
+    await tester.pumpAndSettle();
+    expect(find.text('1'), findsOneWidget);
+  });
+  testWidgets('Sets initial value and then listens to valueNotifier', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/value_updater_test.dart:29]: nit: blank line before this line

File: packages/flutter/lib/src/widgets/value_updater.dart [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+/// {@tool dartpad}
+/// This example shows how to use the [ValueUpdater] widget to trigger an
+/// [AnimatedAlign] animation when a widget is shown for the first time, and
+/// then change the value of the alignment and animate each time a button is
+/// pressed. This is done without needing to create a stateful widget.
+///
+/// ** See code in examples/api/lib/widgets/value_updater/value_updater.1.dart **
+/// {@end-tool}
+
+class ValueUpdater<T> extends StatefulWidget {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/value_updater.dart:44]: Two similar widgets to this come to mind: We have the `ValueListenableBuilder` (afaik, it does not support a separate `initialValue`) and the `TweenAnimationBuilder`, which allows you to _animate_ from a start to an end value (specified as a Tween). This widget feels like a non-animated version of the `TweenAnimationBuilder`? For the envisioned use cases, could one of these two existing widgets be used instead for the same effect? Or does the `ValueUpdater` further simplify things? Maybe link to these similar widgets from the doc and describe the difference in a short sentence (e.g. in a "see also" section)?

**Comment by gspencergoog** [comment]
@goderbauer and I spoke about this IRL, and decided that `TweenAnimationBuilder` is an existing widget that can do everything that this widget can do, so I'm just going to abandon this PR.

**Pull request closed without merging**